### PR TITLE
libgtkflow: split into 3 separate derivations

### PR DESCRIPTION
**Pull request by grindhold**
libgtkflow has of late received a port to gtk4.  as there might be users that only need the gtk3-variant as a dependency for a specific application or the gtk4-variant respectively, i've decided to split the three libraries libgflow libgtkflow3 and libgtkflow4 into their own packages.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by grindhold** [comment]
i realize that this PR didn't trigger any automated testing. any help and instructions on how to do such a split-packets-operation properly would be very appreaciated.

File: pkgs/development/libraries/libgtkflow3/default.nix [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    meson
+    ninja
+    pkg-config
+    gobject-introspection
+  ];
+
+  buildInputs = [
+    gtk3
+    glib
+    libgflow
```
Comment (jtojnar) [inline pkgs/development/libraries/libgtkflow3/default.nix:26]: It does not look to me like it supports linking gflow from a different package. That would require at least changing these two lines to `dependencencies: [dependency('libgflow')]`:

https://notabug.org/grindhold/libgtkflow/src/11773a2fd02dc6d2d2ca6e9e7fdc349d7b0a0548/libgtkflow3/meson.build#L18-L19

But there are other references to the in-tree directory as well:

https://notabug.org/grindhold/libgtkflow/src/11773a2fd02dc6d2d2ca6e9e7fdc349d7b0a0548/libgtkflow3/meson.build#L27

It is also very weird that they use `custom_target` over [`gnome.generate_gir`](https://mesonbuild.com/Gnome-module.html#gnomegenerate_gir).

So until we manage to make it build with `postPatch = ''rm -r libgflow''`, we probably should not it split like this.

So we either need to patch the code to use external libgflow (similarly like we do with [gssdp here](https://github.com/NixOS/nixpkgs/blob/c1487a18d133ff8c15b674130c618c128ba7b4fe/pkgs/development/libraries/gssdp/standalone-tools.patch)), or give up on perfect split (like we do in [libportal](https://github.com/NixOS/nixpkgs/blob/c1487a18d133ff8c15b674130c618c128ba7b4fe/pkgs/development/libraries/libportal/default.nix), where both `libportal-gtk3` and `libportal-gtk4` duplicate the `libportal` library.)

**Comment by jtojnar** [comment]
> i realize that this PR didn't trigger any automated testing

That’s due to the commit message not containing an attribute name. We do not really have a convention for commit messages when splitting packages but comma separated list of attribute names [should work](https://github.com/nixos/ofborg#automatic-building). Or you can trigger it manually with:

@ofborg build libgflow libgtkflow3 libgtkflow4

**Comment by grindhold** [comment]
@jtojnar i updated upstream meson to support what you describe. i really think, users should not be bound to install stuff they don't need. the builds work for me on x86_64.

@ofborg build libgflow libgtkflow3 libgtkflow4

**Comment by grindhold** [comment]
again

@ofborg build libgflow libgtkflow3 libgtkflow4

**Comment by jtojnar** [comment]
Looks like there is still a hidden coupling. Applying the following patch:

```patch
--- a/pkgs/development/libraries/libgtkflow3/default.nix
+++ b/pkgs/development/libraries/libgtkflow3/default.nix
@@ -32,6 +32,14 @@ stdenv.mkDerivation rec {
     "-Denable_gflow=false"
   ];
 
+  postPatch = ''
+    substituteInPlace meson.build \
+      --replace "subdir('libgflow')" "if get_option('enable_gflow')
+      subdir('libgflow')
+    endif"
+    rm -r libgflow
+  '';
+
   meta = with lib; {
     description = "Flow graph widget for GTK 3";
     homepage = "https://notabug.org/grindhold/libgtkflow";
```

I get

```
libgtkflow3/meson.build:26:0: ERROR: Unknown variable "gflow_inc".
```

**Comment by jtojnar** [comment]
You could also move libgflow into `subprojects/` and [declare the dependency object](https://mesonbuild.com/Subprojects.html#a-subproject-example) there. Then you could use a single `dependency(f'libgflow-{gflow_dep_api}', fallback: ['libgflow', 'libgflow_dep'])` to handle both system-wide and in-tree gflow dependency. Unfortunately Meson requires this rigid structure to be able to do that.

Or even further, make all three of the libraries separate subprojects like [GStreamer does](https://gitlab.freedesktop.org/gstreamer/gstreamer).

**Comment by grindhold** [comment]
@ofborg build libgflow libgtkflow3 libgtkflow4

**Comment by jtojnar** [comment]
This looks better now but there are still dependencies:

https://notabug.org/grindhold/libgtkflow/src/dd05b097d33abfa766d648c3d53456f150a17725/libgtkflow4/meson.build#L56
https://notabug.org/grindhold/libgtkflow/src/dd05b097d33abfa766d648c3d53456f150a17725/libgtkflow4/meson.build#L100

Also all these conditionals are no longer needed:

https://notabug.org/grindhold/libgtkflow/src/dd05b097d33abfa766d648c3d53456f150a17725/libgflow/meson.build#L38

**Comment by grindhold** [comment]
@ofborg build libgflow libgtkflow3 libgtkflow4

**Comment by grindhold** [comment]
now this looks promising finally

**Comment by jtojnar** [comment]
Looks good, would be nice to add `outputs = [ "out" "dev" "devdoc" ];` between `version` and `src` to have it include the API docs. (Otherwise the valadoc files will be cleaned up.)

**Comment by jtojnar** [comment]
And would be nice to have the demo in the devdoc output too, see e.g. libadwaita:

https://github.com/NixOS/nixpkgs/blob/ae1d03834ebe4881dca31da399e277657ba33779/pkgs/development/libraries/libadwaita/default.nix#L26

**Comment by grindhold** [comment]
@ofborg build libgflow libgtkflow3 libgtkflow4

**Comment by grindhold** [comment]
i've added the splits to -dev and -devdoc. i'd rather leave the demo app out. it only really makes sense when you have the sourcecode too and you are able to play around with it.  i even provide a nix-shell in the source-repo, so that anyone who wants to play on that level has minimal friction.

**Comment by jtojnar** [comment]
Sure, thanks.

**Pull request merged**